honeybadgermpc/progs/fixedpoint.py

```diff
@@ -0,0 +1,229 @@
+import logging
+import asyncio
+import time
+from honeybadgermpc.mpc import TaskProgramRunner
+from honeybadgermpc.elliptic_curve import Subgroup
+from honeybadgermpc.field import GF
+from honeybadgermpc.preprocessing import (
+    PreProcessedElements as FakePreProcessedElements)
+from honeybadgermpc.progs.mixins.share_arithmetic import (
+    MixinConstants, BeaverMultiply)
+
+
+config = {MixinConstants.MultiplyShare: BeaverMultiply(), }
+
+
+# Fixed Point parameters
+
+F = 32  # The precision (binary bits)
+KAPPA = 32  # Statistical security parameter
+K = 64  # Total number of padding bits (?)
+p = modulus = Subgroup.BLS12_381
+Field = GF(p)
+
+
+# General (non MPC) fixed point functions
+def to_fixed_point_repr(x, f=F):
+    return int(x * 2 ** f)
+
+
+def from_fixed_point_repr(x, k=K, f=F, signed=True):
+    x = x.value
+    if x >= 2 ** (k - 1) and signed:
+        x = -(p - x)
+
+    return float(x) / 2 ** f
+
+
+def binary_repr(x, k):
+    """
+    Convert x to a k-bit representation
+    Least significant bit first
+    """
+    def _binary_repr(v):
+        res = []
+        v = int(v)
+        for i in range(k):
+            res.append(v % 2)
+            v //= 2
+            return res
+
+    assert type(x) is int
+    return _binary_repr(x)
+
+
+# MPC operations for fixed point
+async def random2m(ctx, m):
```

Comment:
why is this async?

Reply:
Was async before since the random bits were generated on the fly. Doesn't need to be async anymore though.

Reply:
I don't understand this, but making this async fails the test :(

Reply:
Although I see no reason why this should be async